Create a Preact component

insberr opened the pull request:
## What is this PR for?
This PR is for making a Preact component by converting the existing React component.

## Tasks
- [x] Switch `react` dependencies over to `preact`
- [x] Reconfigure Typescript, Rollup, and Vite to use `preact`
- [x] Fix imports to use `preact`
- [x] Convert `react` component to `preact`. The hope is not to use `preact/compat`
- [x] Update development Vite app
- [x] Update development `next-app`
- [ ] Test everything and make sure it works

tomosterlund commented on package.json:
```diff
@@ -31,11 +31,11 @@
     "@schedule-x/calendar": "1.50.0",
-    "react": "^16.7.0 || ^17 || ^18",
-    "react-dom": "^16.7.0 || ^17 || ^18"
+    "preact": "^10.22.1"
```
I think we can probably pretty safely go with `"preact": "^10"` to allow all minor versions of the current major.

tomosterlund reviewed (commented):
Really cool! Thanks a lot for contributing with this.

I'll go ahead and add an NPM token for the release workflow next time I'm on a private network (right now I'm working in the train 😄 ), and then we can try to release it once you're done. Also don't hesitate with any questions :)

insberr commented:
Sounds like a plan! And hey I am also on a train right now 👀

insberr commented on rollup.config.js:
```diff
@@ -29,6 +30,14 @@ export default {
   ],
   plugins: [
     external(),
+    alias({
```
I am actually not sure. I guess it would make sense that it is not needed, as long as none of the dependencies in the project use React, which should be the case.

insberr commented:
Sorry for the delays! I have removed the alias for `react` and reduced the usage of `preact/compat`. I am unsure what `createPortal` is, but it is the only thing using `preact/compat`. I also ensured the code builds, fixed the development Vite app to use `preact`, and tested running it.

tomosterlund commented:
Sounds great 👏

I'll have a look today. `createPortal` is used to inject custom components into empty placeholder-divs, to override the markup for things like the events, the modal or parts of the header:

https://schedule-x.dev/docs/calendar/react#custom-components

tomosterlund commented:
I noticed now that I ran it, that we need to solve this error:

![Bildschirmfoto 2024-07-15 um 08 11 31](https://github.com/user-attachments/assets/c9e59f66-0f0a-4028-9359-a3ed895678b0)

Probably in the `useCalendarApp` hook, we need to use `@preact/signals`, since in Preact we can't use hooks outside of the component. I think also we can remove the `useNextCalendarApp`

insberr commented:
Of course! Changing that to a signal should be pretty easy.

I believe the persistence and double rendering comes from the hook being used outside of a component. I have found that sometimes things will have persistence issues and double (or triple) render when a breaking but non-fatal error occurs (like using a hook outside of a component). I'll have to take a look though to be sure though.

insberr commented:
Making it a signal fixes the issue, however, it means only one calendar can be created (all new calendars will be exactly the same)

Having the signal inside the useCalendar function still creates the error and must be outside of it.
